Added `InstructionsPanel` component

BrookJeynes opened the pull request:
## Changes
- Added `InstructionsPanel` component
- Added `Tooltip` component
- Added `Level` struct to hold `TerminalContent` and `InstructionsPanel` content
- Added `Levels` struct to hold all level data and provided a function to get a level
- Fixed bug where user could keep entering input in terminal even though the level had ended
- Added mock first level
- Added ability to swap between levels
- Added `class` prop to components to allow style overrides

## Images
<img width="1476" alt="image" src="https://github.com/BrookJeynes/quack-n-hack/assets/25432120/2b6ce5e3-902e-475b-8b4e-f01c70ff60fb">

**Figure: New instructions component on the left**

BrookJeynes committed:
Added `InstructionsPanel` component
- Added `InstructionsPanel` component
- Added `Tooltip` component
- Fixed bug where user could keep entering input in terminal even
  though the level had ended
- Added mock first level
- Added `class` prop to components to allow style overrides

BrookJeynes committed:
Disabled closure linting in Terminal component

BrookJeynes committed:
Fixed linting errors

miyuushi commented on src/pages/home.rs:
```diff
@@ -21,24 +23,74 @@ pub fn Home(cx: Scope) -> impl IntoView {
 
+    // Callbacks
+    let increment_level = move || {
+        set_level(level() + 1);
```
I would change this to `level.update(|n| *n += 1)` as it's the recommended way to update signals.

miyuushi commented on src/pages/home.rs:
```diff
@@ -21,24 +23,74 @@ pub fn Home(cx: Scope) -> impl IntoView {
+
+    let decrement_level = move || {
+        if level() > 0 {
+            set_level(level() - 1);
```
Same change here, but with a decrement.

miyuushi commented on src/components/terminal.rs:
```diff
@@ -2,13 +2,16 @@ use crate::types::terminal_content::TerminalContent;
 pub fn Terminal<F>(
     cx: Scope,
-    // TODO - Tech Debt: Find a way to pass terminal in as mut
     terminal: TerminalContent,
```
Is this supposed to be reactive/change later on? It seems like `TerminalContent` is only for one level, so I would assume this is updated? I would consider passing a `Signal<TerminalContent>` if this is the case.

miyuushi reviewed (commented):
So, a bit of a suggestion on how you might want to approach the levels: since you already have a `level` signal, what you could do is make a derived signal of type `TerminalContent` that depends on `level`, so for instance you could index from a `Vec<TerminalContent>` with the current `level` value to get the `TerminalContent` of the current level.

BrookJeynes committed:
wip - Fixing terminal state issues

BrookJeynes committed:
Fixed `Terminal` state issue
- Fixed `Terminal` state issue by wrapping the component in a closure.
- Added ability to swap between levels

BrookJeynes commented on src/components/terminal.rs:
```diff
@@ -1,21 +1,24 @@
 
+// TODO - Tech Debt: Find a way to not allow the whole
+// component but rather the single closure
+#[allow(clippy::redundant_closure)]
```
So many items in Leptos rely on closures to update values, this one might just need to go on all components.

BenSimmers reviewed (approved):
lgtm!

BrookJeynes committed:
Added comments to code
- Added comments to code
- Converted level increment/decrement icons to buttons
- Less confetti :(
- Renamed `complete_callback` to `complete_level_callback` for
  readability
